File: haskell/glue/src/Glue/Lib/List/Length.hs
module Glue.Lib.List.Length where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

length :: IR Eval
length = NativeFunc lengthImpl

lengthImpl :: IR Eval -> Eval (IR Eval)
lengthImpl arg = case arg of
    List xs -> pure $ Integer (fromIntegral $ Prelude.length xs)
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/length.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Length function - returns the number of elements in a list
/// Mirrors Haskell Glue.Lib.List.Length.length exactly
Ir length = IrNativeFunc(lengthImpl);

/// Length function implementation
/// Mirrors Haskell Glue.Lib.List.Length.lengthImpl exactly
Eval<Ir> lengthImpl(Ir arg) {
  return switch (arg) {
    IrList(elements: final elements) => Eval.pure(IrInteger(elements.length)),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
